### DISCUSSION TITLE
Build cvc4 1.7 with antlr from ubuntu 20.04

### DISCUSSION THREAD

**Discussion by come-maiz** [cvc5/cvc5#7273] [Q&A] [answered]
Hello!

**Describe the bug**
I would like to build cvc4 1.7 using the [libantlr package from ubuntu 20.04](https://packages.ubuntu.com/focal/libantlr3c-3.4-0). I'm not sure what should I pass to ./configure so it can find it.

I'm getting:
`  Could NOT find ANTLR (missing: ANTLR_INCLUDE_DIR ANTLR_LIBRARIES)`

**cvc5 version/commit**: 1.7
**Operating system**: Ubuntu 20.04

**`configure.sh` options**  
That's the question. This is for building the [solidity snap](https://github.com/ethereum/solidity/blob/7a9f48155b6513d9a3a52f533b6cc88af10a3967/snap/snapcraft.yaml#L57), so I'm using:

`./configure.sh --prefix=$SNAPCRAFT_STAGE/usr`

**`configure.sh` output**  
`  Could NOT find ANTLR (missing: ANTLR_INCLUDE_DIR ANTLR_LIBRARIES)`

The thing is that `./contrib/get-antlr-3.4` is failing on architectures other than amd64. So I thought it would be better to use the package instead of building it. However my cmake knowledge is very limited, I'm not sure what the script that finds antlr is asking for.

I will appreciate any pointers.

pura vida

**Comment by nafur** [comment]
Please note, that the 1.7 release is almost two and a half years old and we strongly encourage everyone to use the latest release (which would be 1.8) or the current master. Is there a specific reason you want 1.7?

As for the actual problem: you can try building ANTLR with `MACHINE_TYPE=X ./contrib/get-antlr-3.4`. What machine type are you trying to build for?

**Comment by 4tXJ7f** [comment]
To add to what @nafur said, please also note that libantlr won't be enough, CVC4/cvc5 additionally require ANTLR 3 itself to generate the actual code for the parser.

As @nafur mentioned, your best option is to use a newer version of CVC4/cvc5 that has support for other architectures such as ARM (and installing ANTLR using the corresponding mechanism from that version).

**Comment by come-maiz** [comment]
Well, I just wanted to do the smallest change possible to update incrementally. Now it's building in 20.04 amd64, next the other architectures, let people test it a while, and then update dependencies. 

The builds for ppc64le and aarch64 are failing with:
```
configure: error: cannot guess build type; you must specify one
```

The builds for armv7l and s390x are failing with:
```
./antlr3config.h:167:16: error: duplicate 'unsigned'
```

Since this seems to be solved in the archive already for all the architectures, I thought using the packages antlr3 and libantlr3c-3.4-0 would be easy.

Of course I can change the plan and try with cvc 1.8. Thanks for your replies! :)

**Reply by 4tXJ7f** [reply]
> `configure: error: cannot guess build type; you must specify one`

This error seems familiar. IIRC they are the reason why we patch `config.guess` and `config.sub` files in the ANTLR source before trying to configure it. See our [CMake file for ANTLR](https://github.com/cvc5/cvc5/blob/master/cmake/FindANTLR3.cmake#L61-L89) to get an idea of what we are doing.

> Since this seems to be solved in the archive already for all the architectures, I thought using the packages antlr3 and libantlr3c-3.4-0 would be easy.

If you do have access to those packages on your distribution, we should be able to use them, otherwise we'd have to fix that. Could you try if newer versions of CVC4/cvc5 are able to detect the system versions? Our build system has evolved quite a bit since CVC4 1.7.

**Comment by come-maiz** [comment]
Sure, I'm trying to build with 1.8.

How do I tell it to use python3?

This doesn't work:

```
PYTHON_EXECUTABLE=python3 ./configure.sh --prefix=$SNAPCRAFT_STAGE/usr
```

It says:

```
Found PythonInterp: /usr/bin/python (found version "2.7.18") 
```

And then fails because it can't find toml, which I'm installing from the `python3-toml` package.

Thank friends!

**Reply by 4tXJ7f** [reply]
There are two options that I can think of: Either pass the `--python3` option to `configure.sh` or configure and compile cvc5 in a Python virtual environment. Hope this helps!

**Comment by come-maiz** [comment]
yes! That worked for finding toml. However it still can't find antlr:

```
Could NOT find ANTLR (missing: ANTLR_INCLUDE_DIR ANTLR_LIBRARIES)
```

I could try building antlr from the 1.8 tag. That's just a little harder to test since I just have an amd64 machine, that's one of the reasons to prefer the already packaged stuff :)

```
# which antlr3
/usr/bin/antlr3
```

```
# ls /usr/lib/x86_64-linux-gnu/libantlr3c-3.4.so.0
/usr/lib/x86_64-linux-gnu/libantlr3c-3.4.so.0
```

**Reply by nafur** [reply]
Note that we also need the antlr3 development files, most probably https://packages.ubuntu.com/focal/libantlr3c-dev. Also, it might be a problem we are simply searching for `libantlr3c.so`, not expecting the `-3.4` suffix: https://github.com/cvc5/cvc5/blob/5247901077efbc7b9016ba35fded7a6ab459a379/cmake/FindANTLR.cmake#L9
An appropriately named library (without the `-3.4` suffix) should also come with the aforementioned package.

**Comment by come-maiz** [comment, marked as answer]
Nice, with the dev package it worked. Thanks a lot for the patient guidance!
